Automattic/sensei
# When lesson is moved between courses, ensure module is valid

Fixes #1806 

Previously, if the lesson was part of a module, and was moved to a course that did not include that module, the lesson would not be removed from that module, which would cause UI inconsistencies.

This commit ensures that the lesson's module is consistent with the available modules in the course that it is moving to.

NOTE: the frontend change in #1891 masks this error, since the JS ensures that a valid module is selected. However, the error is evident when JS is disabled.

## To Test

- Disable Javascript.
- Create two courses with a few lessons.
- Make sure one of the courses has a module while the other one does not and add the lessons to the module.
- Move one of the lessons from that course/module to the other course that doesn't have any modules.
- Go to Lessons > Order Lessons and select the course the lesson was moved to.
- Ensure that the moved lesson appears in the list.

## Other Notes

- ~~I refactored the `save_lesson_module` method. It now only handles authorization and nonce checking, and then calls `set_lesson_module`, which handles the data update logic. I wrote unit tests for `set_lesson_module`, but the simplest way to do so was to make it `public`. I'm not sure whether we want to add this to the plugin's public API?~~ _UPDATE: See https://github.com/Automattic/sensei/pull/1930#issuecomment-335500347_

- Some of the changes here conflicted with some changes in #1891. @pgk please let me know if I messed something up! 😄 

- This file has mixed indentation. After review and acceptance, I will add a commit to convert indentation to tabs. (Looks like some of the indentation changes were automatically fixed by my editor as well).
